### ISSUE TITLE
vcat! or push!(..., columns=:union)

### ISSUE THREAD

**Issue by rapus95** [JuliaData/DataFrames.jl#1982]
Is there a reason why I can't find a good way to add rows, where each comes from a dict with only a subset of the columns set, in-place?

**Comment by bkamins** [comment]
I was thinking about allowing it in `push!` (as this is I guess what you want). I can add it https://github.com/JuliaData/DataFrames.jl/pull/1958 as this is not merged yet.

However, if you are willing please discuss what the target functionality should be. The particular issues are:
* what if the column is missing in added row and the column in a target data frame does not allow `missing` (do we throw an error then - I guess yes)
* what if the target does not have a column that is present in the added row (do we error then? - I guess yes).

The current way to do it is to use `vcat` with `:union` option, and you have to materialize each row as 1-row `DataFrame` (I know it is inefficient and that is why I am OK to consider making `push!` more flexible). The difference is that `push!` will not change target metadata, while `vcat` with `:union` will be still more flexible.

Any opinion on this?

**Comment by rapus95** [comment]
For now I implemented the following as some helper function for my usecase (supporting additional values and assuming that missing is a valid element either way).

```julia
function fillingpush!(df::DataFrame, dict::AbstractDict; additionalvalues...)
    temp = similar(df, 1)
    for col in names(temp)
        temp[col] = get(dict, col, get(additionalvalues, col, missing))
    end
    push!(df, temp)
end
```

> However, if you are willing please discuss what the target functionality should be. The particular issues are:
> 
> * what if the column is missing in added row and the column in a target data frame does not allow `missing` (do we throw an error then - I guess yes)

I'd say yes here aswell.

> * what if the target does not have a column that is present in the added row (do we error then? - I guess yes).

When `:union` is selected I'd act differently depending upon whether the requested column supports Missing:
 * Supports Missing: Add column that has missing for every row of the first DataFrame
 * Does not support Missing: Throw an error as we couldn't keep the type.
 
Thus, as a general rule, I'd say `push!` never changes column types (not even widening them to support missing). Everyone can do that explicitly anyway.
Thus:
```julia
canhold_or_convert(X,Y) #tests, if we can put x::X into a cell of type Y
A,B #be DataFrames
easy = intersect(names(A), names(B))
@assert canhold_or_convert.(types(B[easy]),types(A[easy]))
onlyA = setdiff(names(A),names(B))
@assert canhold_or_convert.(Missing, types(A[onlyA]))
onlyB = setdiff(names(B),names(A))
@assert canhold_or_convert.(Missing, types(B[onlyB]))

addcolumns!(A, types(B,onlyB), onlyB, missing) #add columns with correct types as in B
partialpush!(A, B, easy)
partialpush!(A, B, onlyB)
partialpush!(A, onlyA, size(B,1), missing) #add given number of missings for onlyA columns
```

I hope that describes when I would expect it to fail and when to succeed. And what happens to the types. Just ask if unclear 😄

**Comment by bkamins** [comment]
@nalimilan - do you have an opinion here?

I would say that I would prefer `push!` not to change column types and count of target data frame and throw an error if this would be needed. In such case I would call this option `:subset` rather than `:union`.

**Comment by rapus95** [comment]
@bkamins even better that you have ab etter name for it! That way both types of appending can be implemented and coexist 👍

**Comment by nalimilan** [comment]
Your proposal sounds good. `cols=:union` sounds OK for consistency with `vcat` -- throwing an error if the column doesn't allow for `missing` is natural, so I wouldn't say we need a different name.

**Comment by bkamins** [comment]
I think we need a different name, `:union` in `vcat` means that if a column is present in **any** of the argument then it will be present in the result. Conversly in `push!` the result will contain the columns present in the source `DataFrame`.

However, I think we could add `vcat!` that would work exactly like `vcat` but would in-place mutate the first argument passed that would have to be a `DataFrame`, but consecutive arguments could be either data frames, or objects that are currently are accepted in `push!`. This is a minimal abuse of naming convention but would provide a functionality that is often needed in practice, namely:

you have a collection of `Dict`s or `NamedTuple`s, you want to make a `DataFrame` of them. The problem is that in general they have different sets of columns and different types of elements. Try e.g. converging the following data set https://www.yelp.com/dataset (e.g. the JSON with reviews) into a `DataFrame`. You have to first convert each row into a single-row `DataFrame` and then `vcat` them, which is very inefficient.

Alternatively we could add `:union` option to `push!` as originally proposed that would be flexible (i.e. would be allowed to add columns and mutate column types if needed), but my feeling is that it is better to have a separate function, like `vcat!` for it (if you have a better name then please comment).

**Comment by nalimilan** [comment]
Ah, yes, we could have both `:union` and `:subset` if needed.

What would be the difference between `push!` and `vcat!`?

**Comment by bkamins** [comment]
If we have `:union` in `push!` then `vcat!` is not needed I think. If we agree we want both options in `push!` then I will implement them in #1958. Could you please confirm before I move forward with the implementation? (maybe e.g. we really need only `:union`, as thinking of it now `:subset` is not that useful if we want to allow `:union`)

**Comment by rapus95** [comment]
Well, if you implement `:union` I'd ask to implement `:subset` by enforcing names being subset and then calling `:union`.
As such it clearly is a convenience function.

:+1:

When you say that you implement push!, does that mean that append! will work aswell?

**Comment by rapus95** [comment]
Or as a general question, would it be useful to have some `strict` keyword which decides, whether the set relation (union, subset, intersect etc) will be enforced (throwing an error if not fit) or whether the resulting columns are just a filtered (and optimized) version of the union merge?

**Comment by bkamins** [comment]
Currently `append!` is always strict. I was thinking about changing `push!`. If you want a *flexible* `append!` I mentally thought of using `vcat`.

Now regarding `:union` - I am ok to also have `:subset` that is strict as we discussed. Now as I think of `:union` the only problem is column order - would we allow it to be flexible or not. And if yes, then what order should we use for new columns when we push a `Dict`?

**Comment by nalimilan** [comment]
`push!` and `append!` should probably support the same arguments. Why not allow `:subset` too.

**Comment by bkamins** [comment]
@nalimilan - I have updated #1958 to reflect the proposed changes. If we decide we like it then I will update `append!` likewise in a follow-up PR.

**Comment by bkamins** [comment]
Closed in favor of #2032
